Refactor flags in tracee-rules

grantseltzer opened the pull request:
This addresses some of the points in #499. In particular:

1) Removes the `--trace-file`, and `--stdin-as` flags.
1) Adds a new flag called `--input-trace` that allows the user to specify the input file and input format for events to tracee-rules.

Still need to address:
1) ~Need to add unit tests for the above parsing logic~
1) What should the default options be?
1) ~https://github.com/aquasecurity/tracee/issues/499#issuecomment-771927000~

Signed-off-by: grantseltzer <grantseltzer@gmail.com>

grantseltzer commented:
Also need to address: https://github.com/aquasecurity/tracee/issues/499#issuecomment-771927000

grantseltzer commented:
Thoughts on default options being reading `gob` from `stdin`? 

Also for the `eof` option, should this mean cut the stream/execution once an EOF is reached in the reading file?

itaysk commented on tracee-rules/main.go:
```diff
@@ -16,6 +17,9 @@ func main() {
 		Usage: "A rule engine for Runtime Security",
 		Action: func(c *cli.Context) error {
+			if c.NArg() == 0 && c.NumFlags() == 0 {
+				return errors.New("tracee-rules requires arguments, please use `tracee-rules --help`")
```
isn't it better to just print the help? I think you or Simar suggested that

itaysk commented on tracee-rules/main.go:
```diff
@@ -16,6 +17,9 @@ func main() {
 		Name:  "tracee-rules",
 		Usage: "A rule engine for Runtime Security",
 		Action: func(c *cli.Context) error {
+			if c.NArg() == 0 && c.NumFlags() == 0 {
```
for my understanding, what are both needed?

grantseltzer replied:
They're actually not, this stemmed from me confusing this CLI library with the other popular Go one (Cobra) which does positional argument validation. I'm removing the `NArg()` check. Tracee-rules doesn't take any positional args anyway.

itaysk commented on tracee-rules/output.go:
```diff
@@ -16,6 +16,7 @@ func setupOuput(webhook string) (chan types.Finding, error) {
 	go func() {
 		for res := range out {
 			fmt.Printf("%+v\n", res)
+			//FIXME: Don't do this check everytime there's a finding
```
ok but can you please record this in an issue instead of a comment?

itaysk commented on tracee-rules/input.go:
```diff
@@ -4,46 +4,46 @@ import (
+type inputFormat uint8
+
+const (
+	invalidInputFormat inputFormat = iota + 1
```
curios why `+ 1`?

grantseltzer replied:
Snuck in from a change I made. Removing.

itaysk reviewed (commented):
added some small comments, but the big one is to properly handle EOT flag. I've commented on the issue, let's first make sure you understand the requirement first (this is crucial to how tracee-ebpf->rules work)

grantseltzer commented:
@itaysk I addressed your feedback, can you also verify my usage of `tracee.Event` vs. `types.Event`. I had to rebase and solve the conflict so just want to make sure I got the idea right of where to use each.

yanivagman commented on tracee-rules/input.go:
```diff
@@ -4,47 +4,49 @@ import (
-		close(res)
-	}()
-	return res, nil
+var helpErr = errors.New("help me if you can im feeling down")
```
:-)

itaysk replied:
a better way to implement this would be to return a typed error, and check for the error type, instead of it's string representation.

grantseltzer replied:
I did this based on how the Go standard library does it (i.e. io.EOF in https://golang.org/pkg/io/#pkg-variables). 

Do you think it's necessary to create a new type? We'd then have to have a type switch on return and it'd be effectively doing the same thing.

itaysk replied:
I don't think that's a comparable example, but it's not that important, this is fine as well

yanivagman commented on tracee-rules/input.go:
```diff
@@ -4,47 +4,49 @@ import (
-	_, err := os.Stat(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid Tracee input file: %s", traceeFilePath)
+func setupInputSource(opts *traceeInputOptions) (chan types.Event, error) {
```
Is there something specific to tracee in the opts struct?
If not, better changing the name to `inputOptions`, especially if you change `setupTraceeSource`->`setupInputSource`

grantseltzer replied:
This is a fair point, there's of course value in keeping code more 'general', but my philosophy is that we shouldn't complicate code for the sake of flexibility that we don't need yet. This isn't such a large chunk of code, when we support more input sources we can easily go back and refactor a bit.

yanivagman replied:
Ok, got it.
In that case, I think the name should be changed to `setupTraceeInputSource` as I mentioned above

itaysk replied:
if you do that then same for setup{GOB,JSON}InputSource (add tracee since they accept traceeOptions)

itaysk replied:
@grantseltzer this (and the others) still need to be changed

grantseltzer replied:
Got it. Agreed they need to be changed but this will require a bit more work because  when `parseTraceeInputOptions` is called, the channel of type `chan tracee.Events` can't be assigned to `chan types.Events` (channels of types are themselves their own types). I'm not sure fixing this should be in this PR, I can fix that in a follow up if you'd like.

itaysk replied:
I don't understand what you wrote :) we are just suggesting the rename the function name, shouldn't have any implications. but if you want to leave it, that's fine

grantseltzer replied:
Lol I thought you were responding to https://github.com/aquasecurity/tracee/pull/500#issuecomment-772813353

yanivagman commented on tracee-rules/input.go:
```diff
@@ -66,8 +68,132 @@ func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
 	return res, nil
 }
+
+func parseTraceeInputOptions(inputOptions []string) (*traceeInputOptions, error) {
```
Same question as above - is there something specific to tracee here?
If yes, is it possible to make this function generic?

grantseltzer replied:
_(see comment above)_

yanivagman commented on tracee-rules/input.go:
```diff
@@ -66,8 +68,132 @@ func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
+
+func printHelp() {
+	traceeInputHelp := `
+tracee-rules --input-tracee <key:value>,<key:value> --input-tracee <key:value>
```
To be future proof, I think we should change this flag name to be simply `input`.
Then, we can have something like:
`--input source:key=value`
WDYT?

itaysk replied:
It's unlikely that future input sources will share config options. consider kubernetes audit - it needs a URL, credentials, maybe a kubeconfig file. none of this is relevant to tracee, so I think it's better to keep separate config structs per input source and separate flags

grantseltzer replied:
_(see my comment above)_

grantseltzer commented:
I believe that based on the discussion in #508 this is ready for further review and the final question I have is what the default behavior should be for tracee-rules. I think it should be reading gob format from standard input.

itaysk commented:
@grantseltzer the code in this PR doesn't implement the revised logic suggested in #508 , did you mean to say that it is?

regarding the defaults, if you want to include #508 in this PR, then:

```
--input-tracee file:{/path/to/file,stdin} (default: stdin)
--input-tracee format:{gob/json}
```

grantseltzer commented:
@itaysk I figured I would make that a separate PR but certainly can into this one.

grantseltzer commented:
I'm actually going to open it as a second PR, the two things are related but it's a different part of the repo.

itaysk commented:
> I figured I would make that a separate PR but certainly can into this one.

no that's fine, I just understood it this way because you mentioned the discussion.

so now, for gob you read until EOT but for JSON you don't. I'm not sure this is what we wanted, but we can let it slide because that's changing soon anyway

itaysk commented on tracee-rules/input.go:
```diff
@@ -66,8 +67,118 @@ func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
 			}
 		}
-		f.Close()
+		opts.inputFile.Close()
```
since you have unified stdin and file, you will now close stdin. I'm not sure whether this is good or not

grantseltzer replied:
I think that's fine. At this point in the code we're going to be closing down so we won't need to read from this stdin fd anyway. Does feel a little weird though.

yanivagman replied:
What about the case when there are multiple input sources? Not sure we can assume we're closing down
Also see my answer in #508

itaysk replied:
yes we can't assume we're closing the program but I guess closing the file won't do any harm

itaysk commented on tracee-rules/input.go:
```diff
@@ -66,8 +67,118 @@ func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
+
+'tracee-rules --input-tracee file:./events.json --input-tracee format:json'
+'tracee-rules --input-tracee file:./events.gob --input-tracee format:gob'
+'sudo tracee-ebpf -o format:gob -o option:eot | tracee-rules --input-tracee file:stdin --input-tracee format:gob'
```
remove eot

itaysk commented on tracee-rules/main.go:
```diff
@@ -51,28 +63,24 @@ func main() {
+			},
+			&cli.StringSliceFlag{
+				Name:  "input-tracee",
+				Usage: "specify various key:value pairs for input options from tracee-ebpf, use '--input-tracee help' for more info",
```
may I suggest an alternate text:
`configure tracee-ebpf as input source. see '--input-tracee help' for more info`

itaysk commented on tracee-rules/input.go:
```diff
@@ -66,8 +67,118 @@ func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
+	res := make(chan types.Event)
+	scanner := bufio.NewScanner(opts.inputFile)
+	go func() {
+		for scanner.Scan() {
```
restating what I wrote in a general comment. if we decided to always read until EOT, then this should be the behavior here as well. but since it's about to be removed, we can forgive this one this time

simar7 commented on tracee-rules/input.go:
```diff
@@ -4,70 +4,174 @@ import (
+			if err != nil {
+				log.Printf("invalid json in %s: %v", string(event), err)
+			}
+			res <- tracee.Event(e)
```
```suggestion
			res <- e
```

simar7 commented on tracee-rules/input.go:
```diff
@@ -4,70 +4,174 @@ import (
+	if opts.inputFormat == jsonInputFormat {
+		return setupTraceeJSONInputSource(opts)
 	}
-	f, err := os.Open(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid file: %s", traceeFilePath)
+
+	if opts.inputFormat == gobInputFormat {
+		return setupTraceeGobInputSource(opts)
```
nit: Replace with a switch statement.

grantseltzer replied:
If it's all the same to you, I think this looks cleaner than this:

```

simar7 commented on tracee-rules/input.go:
```diff
@@ -4,70 +4,174 @@ import (
-					continue
+					break
 				} else {
 					log.Printf("Error while decoding event: %v", err)
```
Is it possible for the input file to be closed when we get here? If that happens we will be stuck in a loop forever.

simar7 replied:
(I realize it's not a change of logic here but it's just me looking at this with a fresh set of eyes)

grantseltzer replied:
Hm, that is a good point but i'm not sure how it would?

simar7 commented on tracee-rules/input.go:
```diff
@@ -4,70 +4,174 @@ import (
+	return nil, errors.New("could not set up input source")
+}
+
+func setupTraceeGobInputSource(opts *traceeInputOptions) (chan types.Event, error) {
```
This is mostly a question for myself but why does this simple test program not work?

```
func Test_setupTraceeGobInputSource(t *testing.T) {
	f, err := ioutil.TempFile("", "Test_setupTraceeGobInputSource-*")
	require.NoError(t, err)
	defer func() {
		_ = f.Close()
		_ = os.RemoveAll(f.Name())
	}()

	// write to file
	e := gob.NewEncoder(f)
	event := tracee.Event{
		EventName: "foo",
	}
	require.NoError(t, e.Encode(event))
	
	opts := traceeInputOptions{
		inputFile:   f,
		inputFormat: gobInputFormat,
	}

	res, err := setupTraceeGobInputSource(&opts)
	require.NoError(t, err)
	assert.Equal(t, "foo", (<-res).(tracee.Event).EventName)
}
```

simar7 commented on tracee-rules/input.go:
```diff
@@ -4,70 +4,174 @@ import (
+	return nil, errors.New("could not set up input source")
+}
+
+func setupTraceeGobInputSource(opts *traceeInputOptions) (chan types.Event, error) {
```
Is this function tested somewhere else? Maybe we could write a simple test like:
```
func Test_setupTraceeGobInputSource(t *testing.T) {
	f, err := ioutil.TempFile("", "Test_setupTraceeGobInputSource-*")
	require.NoError(t, err)
	defer func() {
		_ = f.Close()
		_ = os.RemoveAll(f.Name())
	}()

	// write to file
	e := gob.NewEncoder(f)
	event := tracee.Event{
		EventName: "foo",
	}
	require.NoError(t, e.Encode(event))
	f.Seek(0, io.SeekStart)

	opts := traceeInputOptions{
		inputFile:   f,
		inputFormat: gobInputFormat,
	}

	res, err := setupTraceeGobInputSource(&opts)

	require.NoError(t, err)
	assert.Equal(t, "foo", (<-res).(tracee.Event).EventName)
}
```

grantseltzer replied:
In the interest of unblocking Itay I'm going to include this in a follow up PR, but yes you do make a good point, testing these like how you suggested would be a big win.

grantseltzer replied:
Issue created: #517

itaysk reviewed (approved):
please also update the readme accordingly and squash before you merge
